File: 0020.py
class Solution:
    def isValid(self, str):
        stack = []
        self.str = str
        for char in str:
            if char == '{':
                stack.append('{')
            elif char == '[':
                stack.append('[')
            elif char == '(':
                stack.append('(')
            else:
                if stack == []:
                    return False
                if char == '}' and stack[-1] == '{':
                    stack.pop()
                elif char == ']' and stack[-1] == '[':
                    stack.pop()
                elif char == ')' and stack[-1] == '(':
                    stack.pop()
                else:
                    return False
        return stack == []
        

File: src/0020.cpp
#include <iostream>
#include <stack>

using namespace std;

class Solution {
public:
    bool isValid(string s) {
        stack<char> st;
        for(char c : s) {
            switch(c) {
                case '[' : st.push(c); break;
                case '{' : st.push(c); break;
                case '(' : st.push(c); break;
                default : 
                    if(st.empty()) return false;
                    switch (c) {
                    case ']':
                        if(st.top() == '[') {
                            st.pop();
                            break;
                        }else {
                            return false;
                        }
                    case '}':
                        if(st.top() == '{') {
                            st.pop();
                            break;
                        }else {
                            return false;
                        }
                    case ')':
                        if(st.top() == '(') {
                            st.pop();
                            break;
                        }else {
                            return false;
                        }
                    default:
                        break;
                    }
            }
        }
        return st.empty(); 
    }
};
